File: Swift/ValidParentheses.swift
//
//  LeetCode
//  Created by Abdelrahman Nasser
//

class Solution {
    func isValid(_ s: String) -> Bool {
        let charMap: [Character:Character] = [
        "}": "{",
        ")": "(",
        "]": "["
        ]
        var openBrackets = [Character]()
         for char in s {
            let c = charMap[char]
            if c == nil {
                openBrackets.append(char)
            } else {
                if openBrackets.popLast() != c {
                    return false
                }
            }
        }
        return openBrackets.count == 0
    }
}

File: Java/ValidParentheses.java
//
//  LeetCode
//  Created by Abdelrahman Nasser
//

class Solution {
    public boolean isValid(String s) {
        HashMap map = new HashMap();
        map.put('(',')');
        map.put('[',']');
        map.put('{','}');
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {

            char a = s.charAt(i);
            
            if (a == '(' || a == '[' || a == '{') {
                stack.push(a);
            } else if (stack.isEmpty()) {
                return false;
            } else if (map.get(stack.pop()).equals(a)) {
                continue;
            } else return false;
        }
        return stack.isEmpty();
    }
}